src/graph/network/socket_communicator.cc

```diff
@@ -0,0 +1,213 @@
+  if (is_sender) {
+    is_sender_ = true;
+    return InitSender(ip, port);
+  } else {
+    is_sender_ = false;
+    return InitReceiver(ip, port, num_sender, queue_size);
+  }
+}
+
+bool SocketCommunicator::InitSender(const char* ip, int port) {
+  // Sender only has a client socket
+  socket_.resize(1);
+  socket_[0] = new TCPSocket();
+  TCPSocket* client = socket_[0];
+  bool bo = false;
+  int try_count = 0;
+  // Connect to server
+  while (bo == false && try_count < kMaxTryCount) {
+    if (client->Connect(ip, port)) {
+      LOG(INFO) << "Connected to " << ip << ":" << port;
+      return true;
+    } else {
+      LOG(ERROR) << "Cannot connect to " << ip << ":" << port
+                 << ", try again ...";
+      bo = false;
+      try_count++;
+#ifdef _WIN32
+      Sleep(1);
+#else   // !_WIN32
+      sleep(1);
+#endif  // _WIN32
+    }
+  }
+  return false;
+}
+
+bool SocketCommunicator::InitReceiver(const char* ip,
+                                      int port,
+                                      int num_sender,
+                                      int64_t queue_size) {
+  CHECK_GE(num_sender, 1);
+  CHECK_GT(queue_size, 0);
+  // Init message queue
+  num_sender_ = num_sender;
+  queue_size_ = queue_size;
+  queue_ = new MessageQueue(queue_size_, num_sender_);
+  // Init socket, and socket_[0] is the server socket
+  socket_.resize(num_sender+1);
+  thread_.resize(num_sender);
+  socket_[0] = new TCPSocket();
+  TCPSocket* server = socket_[0];
+  server->SetTimeout(kTimeOut * 60 * 1000);  // millsec
+  // Bind socket
+  if (server->Bind(ip, port) == false) {
+    LOG(ERROR) << "Cannot bind to " << ip << ":" << port;
+    return false;
+  }
+  LOG(INFO) << "Bind to " << ip << ":" << port;
+  // Listen
+  if (server->Listen(kMaxConnection) == false) {
+    LOG(ERROR) << "Cannot listen on " << ip << ":" << port;
+    return false;
+  }
+  LOG(INFO) << "Listen on " << ip << ":" << port << ", wait sender connect ...";
+  // Accept all sender sockets
+  std::string accept_ip;
+  int accept_port;
+  for (int i = 1; i <= num_sender_; ++i) {
+    socket_[i] = new TCPSocket();
+    if (server->Accept(socket_[i], &accept_ip, &accept_port) == false) {
+      LOG(ERROR) << "Error on accept socket.";
+      return false;
+    }
+    // new thread for the socket
+    thread_[i-1] = new std::thread(MsgHandler, socket_[i], queue_);
+    LOG(INFO) << "Accept new sender: " << accept_ip << ":" << accept_port;
+  }
+
+  return true;
+}
+
+void SocketCommunicator::MsgHandler(TCPSocket* socket, MessageQueue* queue) {
+  char* buffer = new char[kMaxBufferSize];
+  for (;;) {
+    // First recv the size
+    int64_t received_bytes = 0;
+    int64_t data_size = 0;
+    while (received_bytes < sizeof(int64_t)) {
+      int64_t max_len = sizeof(int64_t) - received_bytes;
+      int64_t tmp = socket->Receive(
+        reinterpret_cast<char*>(&data_size)+received_bytes,
+        max_len);
+      received_bytes += tmp;
+    }
```

Comment:
it seems you use a while loop to send and receive data in a buffer everywhere. why not move the while loop in the Send and Receive method?

Reply:
We don't know the message size inside the Send() and Recv(). So we should design the protocol outside the Send() and Recv().